src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Kafka/KafkaHelper.cs

```diff
@@ -0,0 +1,220 @@
+
+                if (partition is not null)
+                {
+                    tags.Partition = partition.ToString();
+                }
+
+                if (offset is not null)
+                {
+                    tags.Offset = offset.ToString();
+                }
+
+                if (message is not null && message.Timestamp.Type != 0)
+                {
+                    var consumeTime = span.StartTime.UtcDateTime;
+                    var produceTime = message.Timestamp.UtcDateTime;
+                    var queueTimeMs = Math.Max(0, (consumeTime - produceTime).TotalMilliseconds);
+                    // TODO: should this be a metric or tag?
+                    tags.RecordQueueTimeMs = queueTimeMs.ToString("0");
+                }
+
+                if (message is not null && message.Value is null)
+                {
+                    tags.Tombstone = "true";
+                }
+
+                // Consumer spans should always be measured
+                span.SetTag(Tags.Measured, "1");
+
+                tags.SetAnalyticsSampleRate(KafkaConstants.IntegrationId, tracer.Settings, enabledWithGlobalSetting: false);
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Error creating or populating scope.");
+            }
+
+            return scope;
+        }
+
+        internal static void CloseConsumerScope(Tracer tracer)
+        {
+            try
+            {
+                if (!tracer.Settings.IsIntegrationEnabled(KafkaConstants.IntegrationId)
+                    || !tracer.Settings.KafkaCreateConsumerScopeEnabled)
+                {
+                    // integration disabled, skip this trace
+                    return;
+                }
+
+                var activeScope = tracer.ActiveScope;
+                var currentSpan = activeScope?.Span;
+                if (currentSpan?.OperationName != KafkaConstants.ConsumeOperationName)
+                {
+                    // Not currently in a consumer operation
+                    return;
+                }
+
+                // TODO: record end-to-end time?
+                activeScope.Dispose();
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Error closing Kafka consumer scope");
+            }
+        }
+
+        /// <summary>
+        /// Try to inject the prop
+        /// </summary>
+        /// <param name="context">The Span context to propagate</param>
+        /// <param name="message">The duck-typed Kafka Message object</param>
+        /// <typeparam name="TTopicPartitionMarker">The TopicPartition type (used  optimisation purposes)</typeparam>
+        /// <typeparam name="TMessage">The type of the duck-type proxy</typeparam>
+        internal static void TryInjectHeaders<TTopicPartitionMarker, TMessage>(SpanContext context, TMessage message)
+            where TMessage : IMessage
+        {
+            // TODO: Allow disabling globally
```

Comment:
Does this TODO need to be resolved in this PR?

Reply:
I think that can be removed actually - I was initially questioning whether we needed to do that or not, based on the Java tracer https://github.com/DataDog/dd-trace-java/blob/master/dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/KafkaProducerInstrumentation.java#L87

However, I tested with an older broker (pre <0.11.0) with the 1.4.x library and it just silently didn't propagate them, so I think we're safe. I'll remove this comment